core/src/monetaryvalue.ts

```diff
@@ -0,0 +1,173 @@
+  ) {
+    if (typeof colorOrDenominations === "string") {
+      this.denominations = new Map([[colorOrDenominations, amount]]);
+    } else if (colorOrDenominations instanceof Map) {
+      this.denominations = new Map(colorOrDenominations);
+    } else {
+      this.denominations = new Map();
+    }
+  }
+
+  /**
+   * Return a new MonetaryValue that is the sum of the current value and the given value.
+   *
+   * If a denomination exists in both the current and the passed value, their amounts will
+   * be summed together and used as the result.
+   */
+  add(value: MonetaryValue): MonetaryValue;
+
+  /**
+   * Return a new MonetaryValue that is the result of adding the given color and amount
+   * to the current MonetaryValue.
+   *
+   * If the color already exists in the current value, the two amounts will be summed together
+   * and used as the result.
+   */
+  add(color: string, amount: number): MonetaryValue;
+
+  add(
+    colorOrMonetaryValue: string | MonetaryValue,
+    amount: number = 0,
+  ): MonetaryValue {
+    const addend = this.toDenominations(colorOrMonetaryValue, amount);
+    const sum = this.mergeReduce(addend, (a, b) => a + b);
+
+    return new MonetaryValue(sum);
+  }
+
+  /**
+   * Return a new MonetaryValue that is the difference of the current value and the given value.
+   *
+   * If a denomination exists in both the current and the passed value, the passed amount will
+   * be subtracted from the current amount and used as the result.
+   */
+  subtract(value: MonetaryValue): MonetaryValue;
+
+  /**
+   * Return a new MonetaryValue that is the result of subtracting the given color and amount
+   * from the current MonetaryValue.
+   *
+   * If the color already exists in the current value, the passed amount will be subtracted
+   * from the current amount and used as the result.
+   */
+  subtract(color: string, amount: number): MonetaryValue;
+
+  subtract(
+    colorOrMonetaryValue: string | MonetaryValue,
+    amount: number = 0,
+  ): MonetaryValue {
+    const subtrahend = this.toDenominations(colorOrMonetaryValue, amount);
+    const difference = this.mergeReduce(subtrahend, (a, b) => a - b);
+
+    return new MonetaryValue(difference);
+  }
+
+  /**
+   * Return whether or not the current value contains the passed value.
+   *
+   * A MonetaryValue `X` contains another MonetaryValue `Y` if all the constituent colors of `Y`
+   * are also present in `X`, and for each such color, the amount in `X` is greater than the
+   * amount in `Y`.
+   */
+  contains(value: MonetaryValue): boolean {
+    const difference = this.subtract(value);
+    return difference.sign >= 0;
+  }
+
+  /**
+   * Return the sign of the current value.
+   *
+   * The sign is -1 if the current value contains any colors whose amount is negative.
+   * The sign is 0 if the current value contains are no colors whatsoever.
+   * The sign is 1 if the current value contains only colors whose amount is positive.
+   */
```

Comment:
what's this for in the game~?

Reply:
Good question ... I suppose I was imagining something like the following:

```typescript
if (tokens.subtract(cost).sign < 0) {
  // can't afford the card
}
```

But `contains()` is doing the exact same thing anyway. We could make `sign` private? Or just move the `sign` logic directly into `contains()`, though I think keeping them separate is somewhat cleaner since they're doing separate things - one is determining the sign of the current value, and one is making a judgement based on that sign with respect to another value.

Reply:
I'm just curious~ It's OK to keep them.